File: aisearch-js/src/symbols.mjs





import { count_char, lstrip, rstrip } from "./util.mjs";
import { detect_lang } from "./config.mjs";



export function make_symbol(kind, name, line, indent = 0) {
  return {
    kind,
    name,
    line,
    line_end: 0,
    col: 0,
    indent,
    parent: "",
  };
}

export function symbol_to_dict(s) {
  const d = { kind: s.kind, name: s.name, line: s.line };
  if (s.line_end) d.line_end = s.line_end;
  if (s.parent) d.parent = s.parent;
  return d;
}







const _JS_KW = "get|set|if|for|while|switch|catch|return|typeof|new|do|else|await|class|function|with|try|finally|throw|delete|yield|using|lock";

const PYTHON_PATTERNS = [
  ["class", /^(\s*)class\s+(\w+)/, 2],
  ["function", /^(\s*)(?:async\s+)?def\s+(\w+)/, 2],
];

const JAVASCRIPT_PATTERNS = [
  ["class", /^(\s*)(?:export\s+(?:default\s+)?)?class\s+(\w+)/, 2],
  ["function", /^(\s*)(?:export\s+(?:default\s+)?)?(?:async\s+)?function\s+(\w+)/, 2],
  ["function", /^(\s*)(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?\(/, 2],
  ["function", /^(\s*)(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?(?:\([^)]*\)|\w+)\s*=>/, 2],

  ["method", new RegExp(
    "^(\\s*)(?:async\\s+)?(?!(?:" + _JS_KW + ")\\b)(\\w+)\\s*\\([^)]*\\)\\s*[\\{:]"
  ), 2],
];

const TYPESCRIPT_EXTRA = [
  ["interface", /^(\s*)(?:export\s+)?interface\s+(\w+)/, 2],


  ["type", /^(\s*)(?:export\s+)?type\s+(\w+)\s*[=<{]/, 2],
  ["enum", /^(\s*)(?:export\s+)?(?:const\s+)?enum\s+(\w+)/, 2],
];

const GO_PATTERNS = [
  ["function", /^(\s*)func\s+(?:\([^)]+\)\s+)?(\w+)/, 2],
  ["struct", /^(\s*)type\s+(\w+)\s+struct/, 2],
  ["interface", /^(\s*)type\s+(\w+)\s+interface/, 2],
  ["type", /^(\s*)type\s+(\w+)/, 2],
];

const RUST_PATTERNS = [
  ["function", /^(\s*)(?:pub\s+)?(?:async\s+)?(?:const\s+)?fn\s+(\w+)/, 2],
  ["struct", /^(\s*)(?:pub\s+)?struct\s+(\w+)/, 2],
  ["enum", /^(\s*)(?:pub\s+)?enum\s+(\w+)/, 2],
  ["trait", /^(\s*)(?:pub\s+)?trait\s+(\w+)/, 2],


  ["impl", /^(\s*)impl(?:\s*<[^>]*>)?\s+(?:.*?\sfor\s+)?(\w+)\s*(?:<[^>]*>)?\s*(?:\{|where|$)/, 2],
  ["type", /^(\s*)(?:pub\s+)?type\s+(\w+)/, 2],
  ["macro", /^(\s*)(?:pub\s+)?macro_rules!\s+(\w+)/, 2],
];

const JAVA_PATTERNS = [
  ["class", /^\s*(?:public\s+)?(?:abstract\s+)?(?:final\s+)?class\s+(\w+)/, 1],
  ["interface", /^\s*(?:public\s+)?interface\s+(\w+)/, 1],
  ["enum", /^\s*(?:public\s+)?enum\s+(\w+)/, 1],
  ["method", /^\s*(?:public|private|protected)\s+(?:static\s+)?(?:final\s+)?(?:synchronized\s+)?[\w<>\[\],\s]+\s+(\w+)\s*\(/, 1],
];

const CPP_EXTRA = [
  ["class", /^\s*(?:class|struct)\s+(\w+)/, 1],
  ["function", /^\s*[\w~:*&]+\s+(?:[\w:*&]+\s+)*(\w+)\s*\([^)]*\)\s*(?:const\s*)?(?:override\s*)?\{/, 1],
  ["namespace", /^\s*namespace\s+(\w+)/, 1],
];

const RUBY_PATTERNS = [
  ["class", /^\s*class\s+(\w+)/, 1],
  ["module", /^\s*module\s+(\w+)/, 1],
  ["method", /^\s*def\s+(\w+)/, 1],
];

const PHP_PATTERNS = [
  ["class", /^\s*(?:abstract\s+)?class\s+(\w+)/, 1],
  ["interface", /^\s*interface\s+(\w+)/, 1],
  ["trait", /^\s*trait\s+(\w+)/, 1],
  ["function", /^\s*(?:public|protected|private)\s+(?:static\s+)?function\s+(\w+)/, 1],
];

const C_PATTERNS = [
  ["struct", /^\s*struct\s+(\w+)/, 1],
  ["function", /^\s*[\w\*]+\s+[\w\*]+\s+(\w+)\s*\([^)]*\)\s*\{/, 1],
];

const SHELL_PATTERNS = [
  ["function", /^(?:function\s+)?(\w+)\s*\(\s*\)/, 1],
];

export const LANG_PATTERNS = {
  python: PYTHON_PATTERNS,
  javascript: JAVASCRIPT_PATTERNS,
  typescript: JAVASCRIPT_PATTERNS.concat(TYPESCRIPT_EXTRA),
  go: GO_PATTERNS,
  rust: RUST_PATTERNS,
  java: JAVA_PATTERNS,
  ruby: RUBY_PATTERNS,
  php: PHP_PATTERNS,
  c: C_PATTERNS,
  cpp: CPP_EXTRA,
  csharp: JAVA_PATTERNS,
  shell: SHELL_PATTERNS,
  lua: [["function", /^(\s*)(?:local\s+)?function\s+(\w+)/, 2]],
  swift: [
    ["class", /^\s*(?:public\s+)?class\s+(\w+)/, 1],
    ["struct", /^\s*(?:public\s+)?struct\s+(\w+)/, 1],
    ["protocol", /^\s*(?:public\s+)?protocol\s+(\w+)/, 1],
    ["function", /^\s*(?:public\s+)?(?:static\s+)?func\s+(\w+)/, 1],
  ],
  kotlin: JAVA_PATTERNS.concat([
    ["function", /^\s*(?:suspend\s+)?fun\s+(\w+)/, 1],
  ]),
  scala: [
    ["class", /^\s*(?:case\s+)?class\s+(\w+)/, 1],
    ["object", /^\s*object\s+(\w+)/, 1],
    ["trait", /^\s*trait\s+(\w+)/, 1],
    ["function", /^\s*def\s+(\w+)/, 1],
  ],
};

const CLASS_KINDS = new Set([
  "class", "struct", "interface", "trait", "enum", "module", "namespace", "object",
]);



const _compiled_cache = new Map();

function _get_compiled(lang) {
  if (_compiled_cache.has(lang)) return _compiled_cache.get(lang);
  const patterns = LANG_PATTERNS[lang] ?? [];
  const compiled = patterns.map(([kind, regex, group]) => ({ kind, regex, group }));
  _compiled_cache.set(lang, compiled);
  return compiled;
}






export const MAX_SYMBOLS_PER_FILE = 5000;

export function extract_symbols(lines, lang, file_path = "") {
  
  const compiled = _get_compiled(lang);
  if (!compiled.length) return [];

  const symbols = [];
  let current_class = "";

  for (let idx = 0; idx < lines.length; idx++) {
    if (symbols.length >= MAX_SYMBOLS_PER_FILE) break;
    const line = lines[idx];
    const lineno = idx + 1;
    const stripped = rstrip(line);
    if (!stripped) continue;
    const ls = lstrip(stripped);

    if (ls.startsWith("#") || ls.startsWith("//") || ls.startsWith("--")) continue;

    const indent = line.length - lstrip(line).length;

    for (const cp of compiled) {
      const m = cp.regex.exec(stripped);
      if (m) {
        const name = cp.group ? m[cp.group] : cp.kind;
        if (!name) continue;

        const sym = make_symbol(cp.kind, name, lineno, indent);


        if (CLASS_KINDS.has(cp.kind)) {
          current_class = name;
        } else if (indent > 0 && current_class) {
          sym.parent = current_class;
        }

        symbols.push(sym);
        break;
      }
    }
  }


  _calc_ranges(symbols, lines, lang);
  return symbols;
}


export function iter_symbol_decls(lineIter, lang) {
  



  const compiled = _get_compiled(lang);
  if (!compiled.length) return [];
  const symbols = [];
  let current_class = "";
  for (const pair of lineIter) {
    if (symbols.length >= MAX_SYMBOLS_PER_FILE) break;
    const lineno = pair[0];
    const line = pair[1];
    const stripped = rstrip(line);
    if (!stripped) continue;
    const ls = lstrip(stripped);
    if (ls.startsWith("#") || ls.startsWith("//") || ls.startsWith("--")) continue;
    const indent = line.length - lstrip(line).length;
    for (const cp of compiled) {
      const m = cp.regex.exec(stripped);
      if (m) {
        const name = cp.group ? m[cp.group] : cp.kind;
        if (!name) continue;
        const sym = make_symbol(cp.kind, name, lineno, indent);
        if (CLASS_KINDS.has(cp.kind)) {
          current_class = name;
        } else if (indent > 0 && current_class) {
          sym.parent = current_class;
        }
        symbols.push(sym);
        break;
      }
    }
  }
  return symbols;
}


export function compute_symbol_end_from_chunk(chunk, lang, sym) {
  


  const temp = make_symbol(sym.kind, sym.name, 1, sym.indent);
  _calc_ranges([temp], chunk, lang);
  return sym.line - 1 + temp.line_end;
}


export function _calc_ranges(symbols, lines, lang) {
  
  if (!symbols.length) return;

  const brace_langs = new Set([
    "javascript", "typescript", "java", "go", "rust", "c", "cpp",
    "csharp", "swift", "kotlin", "scala", "php", "ruby",
  ]);
  const n = lines.length;

  if (brace_langs.has(lang)) {



    const delta = new Array(n).fill(0);
    const opens = new Array(n).fill(false);
    for (let j = 0; j < n; j++) {
      let d = 0;
      let has = false;
      for (const ch of lines[j]) {
        if (ch === "{") { d += 1; has = true; }
        else if (ch === "}") d -= 1;
      }
      delta[j] = d;
      opens[j] = has;
    }
    const depth = new Array(n + 1).fill(0);
    for (let j = 0; j < n; j++) depth[j + 1] = depth[j] + delta[j];
    const next_open = new Array(n + 1).fill(n);
    for (let j = n - 1; j >= 0; j--) next_open[j] = opens[j] ? j : next_open[j + 1];

    const activate_at = new Map();
    const activated = new Set();
    symbols.forEach((sym, idx) => {
      const start0 = sym.line - 1;
      const act = next_open[start0];
      if (act < n) {
        if (!activate_at.has(act)) activate_at.set(act, []);


        activate_at.get(act).push([idx, depth[start0]]);
        activated.add(idx);
      }
    });

    const end_of = new Map();
    const active = new Map();
    let max_base = null;
    for (let j = 0; j < n; j++) {
      const starts = activate_at.get(j);
      if (starts) {
        for (const [idx, b] of starts) {
          if (!active.has(b)) active.set(b, []);
          active.get(b).push(idx);
          if (max_base === null || b > max_base) max_base = b;
        }
      }
      const d_after = depth[j + 1];

      while (max_base !== null && max_base >= d_after) {
        for (const idx of active.get(max_base) ?? []) end_of.set(idx, j + 1);
        active.delete(max_base);
        max_base = active.size ? Math.max(...active.keys()) : null;
      }
    }
    symbols.forEach((sym, idx) => {
      let e;
      if (end_of.has(idx)) e = end_of.get(idx);
      else if (activated.has(idx)) e = n;
      else e = sym.line - 1;
      sym.line_end = Math.max(e, sym.line);
    });
  }

  for (const sym of symbols) {
    if (brace_langs.has(lang)) {
      continue;
    } else {



      const base_indent = sym.indent;
      const j0 = sym.line - 1;
      let header_end = j0;
      let header_done = false;
      let bal = 0;
      for (let j = j0; j < n; j++) {
        const stripped = rstrip(lines[j]);
        bal += count_char(stripped, "(") - count_char(stripped, ")");
        header_end = j;
        if (bal <= 0 && stripped.endsWith(":")) {
          header_done = true;
          break;
        }
      }
      const body_start = header_done ? header_end + 1 : sym.line;

      let end = sym.line;
      for (let j = body_start; j < n; j++) {
        const l = lines[j];
        if (l.trim() === "") {
          end = j + 1;
          continue;
        }
        const cur_indent = l.length - lstrip(l).length;
        if (cur_indent > base_indent) {
          end = j + 1;
        } else {
          break;
        }
      }

      while (end > sym.line && lines[end - 1].trim() === "") end -= 1;
      sym.line_end = Math.max(end, sym.line);
    }
  }
}


export function find_symbol_by_name(lines, lang, name, partial = false) {
  
  const syms = extract_symbols(lines, lang);
  for (const s of syms) {
    if (partial) {

      if (s.name.toLowerCase().includes(name.toLowerCase())) {
        return s;
      }
    } else if (s.name === name) {
      return s;
    }
  }
  return null;
}


export function find_containing_symbol(lines, lang, target_line) {
  
  const syms = extract_symbols(lines, lang);
  let best = null;
  let best_size = Infinity;

  for (const s of syms) {
    if (s.line <= target_line && target_line <= s.line_end) {
      const size = s.line_end - s.line;
      if (size < best_size) {
        best = s;
        best_size = size;
      }
    }
  }
  return best;
}


export function extract_imports(lines, lang) {
  
  const imports = [];


  if (lang === "go") {
    let in_block = false;
    for (const line of lines) {
      const s = line.trim();
      if (in_block) {
        if (s.startsWith(")")) {
          in_block = false;
        } else {
          const m = /^(?:(\w+)\s+)?"([^"]+)"/.exec(s);
          if (m) {
            imports.push(m[1] ? m[1] + " " + m[2] : m[2]);
          }
        }
        continue;
      }
      let m = /^import\s+\($/.exec(s);
      if (m) {
        in_block = true;
        continue;
      }
      m = /^import\s+"([^"]+)"/.exec(s);
      if (m) {
        imports.push(m[1]);
      }
    }
    return imports;
  }

  const patterns = {
    python: /^\s*(?:from\s+\S+\s+)?import\s+(.+)/,
    javascript: /^\s*import\s+.*from\s+['"](.+)['"]/,
    typescript: /^\s*import\s+.*from\s+['"](.+)['"]/,
    rust: /^\s*use\s+(.+)/,
    java: /^\s*import\s+([\w.]+)/,
  };
  const pat = patterns[lang];
  if (!pat) return imports;
  for (const line of lines) {
    const m = pat.exec(rstrip(line));
    if (m) {
      const raw = m[1] ? m[1].trim() : line.trim();
      imports.push(raw);
    }
  }
  return imports;
}

File: aisearch/symbols.py
"""
正则驱动的符号提取 —— 不需要 LSP，开箱即用。
支持 tree-sitter 可选增强（未来扩展，当前以正则为主，零依赖）。
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional

from .config import detect_lang



MAX_SYMBOLS_PER_FILE = 5000



@dataclass
class Symbol:
    kind: str
    name: str
    line: int
    line_end: int = 0
    col: int = 0
    indent: int = 0
    parent: str = ""

    def to_dict(self) -> dict:
        d: dict = {"kind": self.kind, "name": self.name, "line": self.line}
        if self.line_end:
            d["line_end"] = self.line_end
        if self.parent:
            d["parent"] = self.parent
        return d




_LangPatterns = list[tuple[str, str, int]]

PYTHON_PATTERNS: _LangPatterns = [
    ("class",     r"^(\s*)class\s+(\w+)", 2),
    ("function",  r"^(\s*)(?:async\s+)?def\s+(\w+)", 2),
]


_JS_KW = r"get|set|if|for|while|switch|catch|return|typeof|new|do|else|await|class|function|with|try|finally|throw|delete|yield|using|lock"

JAVASCRIPT_PATTERNS: _LangPatterns = [
    ("class",    r"^(\s*)(?:export\s+(?:default\s+)?)?class\s+(\w+)", 2),
    ("function", r"^(\s*)(?:export\s+(?:default\s+)?)?(?:async\s+)?function\s+(\w+)", 2),
    ("function", r"^(\s*)(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?\(", 2),
    ("function", r"^(\s*)(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?(?:\([^)]*\)|\w+)\s*=>", 2),


    ("method",   r"^(\s*)(?:async\s+)?(?!(?:" + _JS_KW + r")\b)(\w+)\s*\([^)]*\)\s*[\{:]", 2),
]

TYPESCRIPT_EXTRA: _LangPatterns = [
    ("interface", r"^(\s*)(?:export\s+)?interface\s+(\w+)", 2),


    ("type",      r"^(\s*)(?:export\s+)?type\s+(\w+)\s*[=<{]", 2),
    ("enum",      r"^(\s*)(?:export\s+)?(?:const\s+)?enum\s+(\w+)", 2),
]

GO_PATTERNS: _LangPatterns = [
    ("function",  r"^(\s*)func\s+(?:\([^)]+\)\s+)?(\w+)", 2),
    ("struct",    r"^(\s*)type\s+(\w+)\s+struct", 2),
    ("interface", r"^(\s*)type\s+(\w+)\s+interface", 2),
    ("type",      r"^(\s*)type\s+(\w+)", 2),
]

RUST_PATTERNS: _LangPatterns = [
    ("function",  r"^(\s*)(?:pub\s+)?(?:async\s+)?(?:const\s+)?fn\s+(\w+)", 2),
    ("struct",    r"^(\s*)(?:pub\s+)?struct\s+(\w+)", 2),
    ("enum",      r"^(\s*)(?:pub\s+)?enum\s+(\w+)", 2),
    ("trait",     r"^(\s*)(?:pub\s+)?trait\s+(\w+)", 2),



    ("impl",      r"^(\s*)impl(?:\s*<[^>]*>)?\s+(?:.*?\sfor\s+)?(\w+)\s*(?:<[^>]*>)?\s*(?:\{|where|$)", 2),
    ("type",      r"^(\s*)(?:pub\s+)?type\s+(\w+)", 2),
    ("macro",     r"^(\s*)(?:pub\s+)?macro_rules!\s+(\w+)", 2),
]

JAVA_PATTERNS: _LangPatterns = [
    ("class",     r"^\s*(?:public\s+)?(?:abstract\s+)?(?:final\s+)?class\s+(\w+)", 1),
    ("interface", r"^\s*(?:public\s+)?interface\s+(\w+)", 1),
    ("enum",      r"^\s*(?:public\s+)?enum\s+(\w+)", 1),
    ("method",    r"^\s*(?:public|private|protected)\s+(?:static\s+)?(?:final\s+)?(?:synchronized\s+)?[\w<>\[\],\s]+\s+(\w+)\s*\(", 1),
]

CPP_EXTRA: _LangPatterns = [
    ("class",     r"^\s*(?:class|struct)\s+(\w+)", 1),
    ("function",  r"^\s*[\w~:*&]+\s+(?:[\w:*&]+\s+)*(\w+)\s*\([^)]*\)\s*(?:const\s*)?(?:override\s*)?\{", 1),
    ("namespace", r"^\s*namespace\s+(\w+)", 1),
]

RUBY_PATTERNS: _LangPatterns = [
    ("class",   r"^\s*class\s+(\w+)", 1),
    ("module",  r"^\s*module\s+(\w+)", 1),
    ("method",  r"^\s*def\s+(\w+)", 1),
]

PHP_PATTERNS: _LangPatterns = [
    ("class",     r"^\s*(?:abstract\s+)?class\s+(\w+)", 1),
    ("interface", r"^\s*interface\s+(\w+)", 1),
    ("trait",     r"^\s*trait\s+(\w+)", 1),
    ("function",  r"^\s*(?:public|protected|private)\s+(?:static\s+)?function\s+(\w+)", 1),
]

C_PATTERNS: _LangPatterns = [
    ("struct",   r"^\s*struct\s+(\w+)", 1),
    ("function", r"^\s*[\w\*]+\s+[\w\*]+\s+(\w+)\s*\([^)]*\)\s*\{", 1),
]

SHELL_PATTERNS: _LangPatterns = [
    ("function", r"^(?:function\s+)?(\w+)\s*\(\s*\)", 1),
]

LANG_PATTERNS: dict[str, _LangPatterns] = {
    "python": PYTHON_PATTERNS,
    "javascript": JAVASCRIPT_PATTERNS,
    "typescript": JAVASCRIPT_PATTERNS + TYPESCRIPT_EXTRA,
    "go": GO_PATTERNS,
    "rust": RUST_PATTERNS,
    "java": JAVA_PATTERNS,
    "ruby": RUBY_PATTERNS,
    "php": PHP_PATTERNS,
    "c": C_PATTERNS,
    "cpp": CPP_EXTRA,
    "csharp": JAVA_PATTERNS,
    "shell": SHELL_PATTERNS,
    "lua": [("function", r"^(\s*)(?:local\s+)?function\s+(\w+)", 2)],
    "swift": [
        ("class",    r"^\s*(?:public\s+)?class\s+(\w+)", 1),
        ("struct",   r"^\s*(?:public\s+)?struct\s+(\w+)", 1),
        ("protocol", r"^\s*(?:public\s+)?protocol\s+(\w+)", 1),
        ("function", r"^\s*(?:public\s+)?(?:static\s+)?func\s+(\w+)", 1),
    ],
    "kotlin": JAVA_PATTERNS + [
        ("function", r"^\s*(?:suspend\s+)?fun\s+(\w+)", 1),
    ],
    "scala": [
        ("class",     r"^\s*(?:case\s+)?class\s+(\w+)", 1),
        ("object",    r"^\s*object\s+(\w+)", 1),
        ("trait",     r"^\s*trait\s+(\w+)", 1),
        ("function",  r"^\s*def\s+(\w+)", 1),
    ],
}




@dataclass
class _CompiledPattern:
    kind: str
    regex: re.Pattern
    group: int

_compiled_cache: dict[str, list[_CompiledPattern]] = {}

def _get_compiled(lang: str) -> list[_CompiledPattern]:
    if lang in _compiled_cache:
        return _compiled_cache[lang]
    patterns = LANG_PATTERNS.get(lang, [])
    compiled = []
    for kind, pat, grp in patterns:
        compiled.append(_CompiledPattern(kind, re.compile(pat), grp))
    _compiled_cache[lang] = compiled
    return compiled




def extract_symbols(
    lines: list[str],
    lang: str,
    file_path: str = "",
) -> list[Symbol]:
    """从源代码行列表中提取所有符号。"""
    compiled = _get_compiled(lang)
    if not compiled:
        return []

    symbols: list[Symbol] = []
    current_class = ""

    for idx, line in enumerate(lines):
        if len(symbols) >= MAX_SYMBOLS_PER_FILE:
            break
        lineno = idx + 1
        stripped = line.rstrip()
        if not stripped:
            continue
        ls = stripped.lstrip()

        if ls.startswith("#") or ls.startswith("//") or ls.startswith("--"):
            continue

        indent = len(line) - len(line.lstrip())

        for cp in compiled:
            m = cp.regex.match(stripped)
            if m:
                name = m.group(cp.group) if cp.group else cp.kind
                if not name:
                    continue

                sym = Symbol(
                    kind=cp.kind,
                    name=name,
                    line=lineno,
                    indent=indent,
                )


                if cp.kind in ("class", "struct", "interface", "trait", "enum", "module", "namespace", "object"):
                    current_class = name
                elif indent > 0 and current_class:
                    sym.parent = current_class

                symbols.append(sym)
                break


    _calc_ranges(symbols, lines, lang)

    return symbols


def iter_symbol_decls(line_iter, lang, file_path: str = "", max_symbols: int = MAX_SYMBOLS_PER_FILE):
    """
    流式扫描符号声明（line_end 置 0，不计算范围），用于大文件避免整文件载入内存。
    line_iter 产出 (lineno, line) 元组。
    """
    compiled = _get_compiled(lang)
    if not compiled:
        return []
    symbols: list[Symbol] = []
    current_class = ""
    for lineno, line in line_iter:
        if len(symbols) >= max_symbols:
            break
        stripped = line.rstrip()
        if not stripped:
            continue
        ls = stripped.lstrip()
        if ls.startswith("#") or ls.startswith("//") or ls.startswith("--"):
            continue
        indent = len(line) - len(line.lstrip())
        for cp in compiled:
            m = cp.regex.match(stripped)
            if m:
                name = m.group(cp.group) if cp.group else cp.kind
                if not name:
                    continue
                sym = Symbol(kind=cp.kind, name=name, line=lineno, indent=indent)
                if cp.kind in ("class", "struct", "interface", "trait", "enum", "module", "namespace", "object"):
                    current_class = name
                elif indent > 0 and current_class:
                    sym.parent = current_class
                symbols.append(sym)
                break
    return symbols


def compute_symbol_end_from_chunk(chunk: list[str], lang: str, sym: Symbol) -> int:
    """
    在从 sym.line 起头的行块上复用 _calc_ranges 推算结束行，并映射回原文行号。
    当块被截断（line_end == len(chunk)）时返回截断处，调用方应再按 total 收敛。
    """
    temp = Symbol(kind=sym.kind, name=sym.name, line=1, indent=sym.indent)
    _calc_ranges([temp], chunk, lang)
    return sym.line - 1 + temp.line_end


def _calc_ranges(symbols: list[Symbol], lines: list[str], lang: str):
    """推算每个符号的结束行：花括号语言按配平计数，缩进语言按缩进扫描。"""
    if not symbols:
        return

    brace_langs = {
        "javascript", "typescript", "java", "go", "rust", "c", "cpp",
        "csharp", "swift", "kotlin", "scala", "php", "ruby",
    }
    n = len(lines)

    if lang in brace_langs:





        delta = [0] * n
        opens = [False] * n
        for j in range(n):
            line = lines[j]
            d = 0
            has = False
            for ch in line:
                if ch == '{':
                    d += 1
                    has = True
                elif ch == '}':
                    d -= 1
            delta[j] = d
            opens[j] = has
        depth = [0] * (n + 1)
        for j in range(n):
            depth[j + 1] = depth[j] + delta[j]

        next_open = [n] * (n + 1)
        for j in range(n - 1, -1, -1):
            next_open[j] = j if opens[j] else next_open[j + 1]







        activate_at: dict[int, list[tuple[int, int]]] = {}
        activated: set[int] = set()
        for idx, sym in enumerate(symbols):
            start0 = sym.line - 1
            act = next_open[start0]
            if act < n:
                activate_at.setdefault(act, []).append((idx, depth[start0]))
                activated.add(idx)

        end_of: dict[int, int] = {}
        active: dict[int, list[int]] = {}
        max_base: Optional[int] = None
        for j in range(n):
            for idx, b in activate_at.get(j, ()):
                active.setdefault(b, []).append(idx)
                if max_base is None or b > max_base:
                    max_base = b
            d_after = depth[j + 1]

            while max_base is not None and max_base >= d_after:
                for idx in active.pop(max_base, ()):
                    end_of[idx] = j + 1
                max_base = max(active) if active else None

        for idx, sym in enumerate(symbols):
            if idx in end_of:
                e = end_of[idx]
            elif idx in activated:
                e = n
            else:
                e = sym.line - 1
            sym.line_end = max(e, sym.line)
    else:
        for sym in symbols:



            base_indent = sym.indent
            j0 = sym.line - 1
            header_end = j0
            header_done = False
            bal = 0
            for j in range(j0, n):
                stripped = lines[j].rstrip()
                bal += stripped.count("(") - stripped.count(")")
                header_end = j
                if bal <= 0 and stripped.endswith(":"):
                    header_done = True
                    break
            body_start = (header_end + 1) if header_done else sym.line

            end = sym.line
            for j in range(body_start, n):
                l = lines[j]
                if l.strip() == '':
                    end = j + 1
                    continue
                cur_indent = len(l) - len(l.lstrip())
                if cur_indent > base_indent:
                    end = j + 1
                else:
                    break

            while end > sym.line and lines[end - 1].strip() == "":
                end -= 1
            sym.line_end = max(end, sym.line)


def find_symbol_by_name(
    lines: list[str],
    lang: str,
    name: str,
    partial: bool = False,
) -> Optional[Symbol]:
    """查找特定名称的符号（partial=True 时支持子串匹配）。"""
    syms = extract_symbols(lines, lang)
    for s in syms:
        if partial:
            if name.lower() in s.name.lower():
                return s
        elif s.name == name:
            return s
    return None


def find_containing_symbol(
    lines: list[str],
    lang: str,
    target_line: int,
) -> Optional[Symbol]:
    """查找包含 target_line 的最小范围符号。"""
    syms = extract_symbols(lines, lang)
    best: Optional[Symbol] = None
    best_size = float("inf")

    for s in syms:
        if s.line <= target_line <= s.line_end:
            size = s.line_end - s.line
            if size < best_size:
                best = s
                best_size = size
    return best


def extract_imports(lines: list[str], lang: str) -> list[str]:
    """提取 import 语句。"""
    imports: list[str] = []






    if lang == "go":
        in_block = False
        for line in lines:
            s = line.strip()
            if in_block:
                if s.startswith(")"):
                    in_block = False
                else:
                    m = re.match(r'(?:(\w+)\s+)?"([^"]+)"', s)
                    if m:
                        imports.append(m.group(1) + " " + m.group(2) if m.group(1) else m.group(2))
                continue
            m = re.match(r'import\s+\($', s)
            if m:
                in_block = True
                continue
            m = re.match(r'import\s+"([^"]+)"', s)
            if m:
                imports.append(m.group(1))
        return imports

    patterns = {
        "python": re.compile(r"^\s*(?:from\s+\S+\s+)?import\s+(.+)"),
        "javascript": re.compile(r"""^\s*import\s+.*from\s+['"](.+)['"]"""),
        "typescript": re.compile(r"""^\s*import\s+.*from\s+['"](.+)['"]"""),
        "rust": re.compile(r"^\s*use\s+(.+)"),
        "java": re.compile(r"^\s*import\s+([\w.]+)"),
    }
    pat = patterns.get(lang)
    if not pat:
        return imports
    for line in lines:
        m = pat.match(line.rstrip())
        if m:
            raw = m.group(1).strip() if m.group(1) else line.strip()
            imports.append(raw)
    return imports
